### DISCUSSION TITLE
Browsing plugin model / API that would allow real-time information access & interaction

### DISCUSSION THREAD

**Discussion by classicjazz** [ggml-org/llama.cpp#465] [Ideas]
Services like GPT4, Bing, etc. now support real-time information access (and actions) beyond the trained foundational model.

The initial idea is to enable llama.cpp to support simple, real-time information services such as weather, sports stats, stock quotes, translation engines, curated/authoritative data sets, etc. (e.g. IMDB, Wolfram, Google Finance). These could be independently maintained, downloaded and manually installed prior to runtime by placing them in a plugin folder.

Beyond web-based services, a plugin could be written for online email inbox or calendar access using the appropriate protocols.

Eventually, trusted services may be permitted to perform constrained actions on behalf of the user (e.g. document generation, to do lists, note taking, etc). The security model would limited not just to user installed plugins but also specifically authorized actions for those plugins. 

Two key advantages of this are user/enterprise privacy and security, as contrasted with doing the same using an online chat service.

For an overview of the possibilities, see [ChatGPT plugins overview](https://openai.com/blog/chatgpt-plugins) and [ChatGPT plugin documentation](https://platform.openai.com/docs/plugins/introduction)

**Comment by ianscrivener** [comment]
Cannot this functionality currently be achieved using say langchain, BabyAGI, SuperAGI etc. I might be missing something - but I do not see that this core to the llama.cpp roadmap
